presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBuilder.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static java.util.Base64.getDecoder;
+
+public class ArrayBuilder
+        extends AbstractVariableWidthBlock
+        implements BlockBuilder
+{
+    // TODO: There might be a better way to exchange binary coded expression
+    private static final int ENCODING_BASE64_FLAG = 0x5F5F5F5F;
+
+    private final BlockBuilderStatus builderStatus;
+    private final VariableWidthBlockBuilder builder;
+
+    public static Block of(Object... values)
+    {
+        return new ArrayBuilder().addAll(values).build();
+    }
+
+    public static Slice sliceOf(Object... values)
+    {
+        return new ArrayBuilder().addAll(values).toSlice();
+    }
+
+    public static ArrayBuilder builder()
+    {
+        return new ArrayBuilder();
+    }
+
+    public static ArrayBuilder builder(int expectedSize)
+    {
+        return new ArrayBuilder(expectedSize);
+    }
+
+    private ArrayBuilder()
+    {
+        this(1024);
+    }
+
+    private ArrayBuilder(int expectedSize)
+    {
+        this(new BlockBuilderStatus(expectedSize, expectedSize));
+    }
+
+    public ArrayBuilder(BlockBuilderStatus builderStatus)
+    {
+        this.builderStatus = builderStatus;
+        this.builder = new VariableWidthBlockBuilder(builderStatus);
+    }
+
+    private static SliceInput getBinaryInput(SliceInput input)
+    {
+        int flags = input.readInt();
+        if ((flags & ENCODING_BASE64_FLAG) == ENCODING_BASE64_FLAG) {
```

Comment:
It has something to do with SliceLiteral, https://github.com/treasure-data/presto/blob/array_map_encoding/presto-parser/src/main/java/com/facebook/presto/sql/tree/SliceLiteral.java

When expression is passed to workers after it is simplified, the simplified expression is formatted as string. For  the complex types, I need a way to deliver VariableWidthBlock encoded bytes as Expression. But I guess there might be a better way to do it.